cep47-logic/src/lib.rs

```diff
@@ -80,120 +86,151 @@ pub trait CEP47Contract<Storage: CEP47Storage>: WithStorage<Storage> {
-        let paused = self.storage().is_paused();
-        if paused {
-            return Err(Error::PermissionDenied);
-        }
-        // 1. Load tokens owned by the sender.
-        let mut sender_tokens = self.storage().get_tokens(sender.clone());
-        // 2. Assert that token_id is in sender_tokens.
-        if !sender_tokens.contains(&token_id) {
-            return Err(Error::PermissionDenied);
-        }
-        // 3. Remove token_id from sender_tokens.
-        sender_tokens.retain(|x| x.clone() != token_id);
-        self.storage_mut().set_tokens(sender, sender_tokens);
-
-        // 4. Add token_id to the recipient tokens
-        let mut recipient_tokens = self.storage().get_tokens(recipient.clone());
-        recipient_tokens.push(token_id);
-        self.storage_mut().set_tokens(recipient, recipient_tokens);
-        Ok(())
+        self.transfer_many_tokens(sender, recipient, &vec![token_id.clone()])
     }
 
     fn transfer_many_tokens(
         &mut self,
-        sender: PublicKey,
-        recipient: PublicKey,
-        token_ids: Vec<TokenId>,
+        sender: &Key,
+        recipient: &Key,
+        token_ids: &Vec<TokenId>,
     ) -> Result<(), Error> {
-        let paused = self.storage().is_paused();
-        if paused {
+        if self.is_paused() {
             return Err(Error::PermissionDenied);
         }
-        let mut sender_tokens = self.storage().get_tokens(sender.clone());
-        println!("Sender tokens: {:?}", sender_tokens);
+        let mut sender_tokens = self.storage().get_tokens(&sender);
 
         for token_id in token_ids.iter() {
             if !sender_tokens.contains(token_id) {
                 return Err(Error::PermissionDenied);
             }
-            sender_tokens.retain(|x| x.clone() != token_id.clone());
+            sender_tokens.retain(|x| x != token_id);
         }
-        let mut recipient_tokens = self.storage().get_tokens(recipient.clone());
+        let mut recipient_tokens = self.storage().get_tokens(&recipient);
         recipient_tokens.append(&mut token_ids.clone());
-        self.storage_mut().set_tokens(sender, sender_tokens);
-        self.storage_mut().set_tokens(recipient, recipient_tokens);
+        self.storage_mut().set_tokens(&sender, sender_tokens);
+        self.storage_mut().set_tokens(&recipient, recipient_tokens);
+
+        // Emit transfer event.
+        let package = self.storage().contact_package_hash();
+        self.storage_mut().emit(CEP47Event::Transfer {
+            sender: sender.clone(),
+            recipient: recipient.clone(),
+            token_ids: token_ids.clone(),
+        });
         Ok(())
     }
 
-    fn transfer_all_tokens(
-        &mut self,
-        sender: PublicKey,
-        recipient: PublicKey,
-    ) -> Result<(), Error> {
-        let paused = self.storage().is_paused();
-        if paused {
+    fn transfer_all_tokens(&mut self, sender: &Key, recipient: &Key) -> Result<(), Error> {
+        if self.is_paused() {
```

Comment:
Why do we only check `paused` state in transfers and not mint, burn, or other entrypoints?

Reply:
Because only transfers are functions avaliable to token holders. Burns and mints are admin only.